### ISSUE TITLE
Exceptions thrown in InferType before doing reverse inference

### ISSUE THREAD

**Issue by anirudh2290** [apache/incubator-mxnet#16757]
## Description
Some operators are missing logic for reverse inference: inferring the type of inputs from outputs. Instead, exception is thrown if input dtype is -1.

Also, InferType logic of operators should throw an exception only if both input and output are defined and incompatible with each other. For example. if inputs to an op are -1 and output is -1, the infertype logic should return false, instead of throwing an exception. It should throw an exception only if the types are incompatible: for example both input types and output types are inferred but input types are not compatible with output types.

The above point is important because the InferType logic does both forward and backward inference. So, sometimes output and input can both be -1 in forward inference but in the backward inference both these values can be inferred.

- [ ] fft
- [ ] ifft
- [ ] deformable_convolution
- [ ] multi_sum_sq
- [ ] SequenceReverse
- [ ] Convolution_v1
- [ ] RNN
- [ ] UpSampling
- [ ] BatchNorm
- [ ] Deconvolution
- [ ] LRN
- [ ] Convolution
- [ ] Embedding
- [ ] split_v2
- [ ] SVMOutput
- [ ] SequenceMask
- [ ] SoftmaxOutput
- [ ] BatchNorm_v1

The existing issues with ops was brought up @ptrendx in https://github.com/apache/incubator-mxnet/pull/16748

**Comment by samskalicky** [comment]
@anirudh2290 Can we make this a feature request and change title to "Improve InferTYpe ..." instead of a bug?

**Comment by anirudh2290** [comment]
It's a bug

**Comment by anirudh2290** [comment]
un-assigning myself since  i am not working on this currently. if someone is interested, please go ahead.
